libraries/AP_CheckFirmware/AP_CheckFirmware.cpp

```diff
@@ -47,7 +78,54 @@ check_fw_result_t check_good_firmware(void)
     if (crc1 != ad->image_crc1 || crc2 != ad->image_crc2) {
         return check_fw_result_t::FAIL_REASON_BAD_CRC;
     }
-    return check_fw_result_t::CHECK_FW_OK;
+
+    check_fw_result_t ret = check_fw_result_t::CHECK_FW_OK;
+
+#if AP_SIGNED_FIRMWARE
+    uint8_t hash[32];
+    int err = wc_ecc_init(&publicKey);
+    if (err != MP_OKAY) {
+        wc_ecc_free(&publicKey);
+        uprintf("Failed to init public key in bootloader\r\n");
+        return check_fw_result_t::FAIL_REASON_BAD_PUBLIC_KEY;
+    }
+
+    // verify signature
+    wc_InitSha256(&sha);
+    wc_Sha256Update(&sha, flash, len1);
+    wc_Sha256Update(&sha, (const uint8_t *)&ad->version_major, len2);
+    wc_Sha256Final(&sha, hash);
+
+    // do ecc verify
+    for (auto public_key : public_keys.public_key) {
```

Comment:
Maybe I'm reading this wrong, but won't it pass if there are no keys at all in the image? Surely it needs to find at least one valid key to be CHECK_FW_OK? I'm thinking ret need to be initialized to whatever the value fr NOTOK is, perhaps?